Split jsp and jsp gui

clalancette opened the pull request:
This PR splits joint_state_publisher into two packages: `joint_state_publisher` and `joint_state_publisher_gui`.  The `joint_state_publisher` package contains both code to be imported (the `JointStatePublisher` class) along with a script that just runs the basic, non-GUI functionality of `JointStatePublisher`.  The `joint_state_publisher_gui` package contains the PyQt UI that allows one to control the movable joint via a UI, and depends on the `joint_state_publisher` package and `JointStatePublisher` class.

Doing this is desirable for a few reasons, mainly to slim down joint_state_publisher and remove the Qt dependencies from the `robot` metapackage (see https://github.com/ros/metapackages/issues/28 for details).

I am unsure where to target this PR.  This is definitely a breaking change in that users who just install/launch `joint_state_publisher` will no longer have GUI functionality available.  So I would usually target noetic with this.  However, I think that removing the GUI dependencies from the `robot` metapackage in kinetic and melodic is important, so I'm currently targeting this at kinetic-devel.  I'm definitely open to discussing this further.

This also fixes #21 .

clalancette committed:
Update the README.md to have topics and parameters in use.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Remove GUI parts from joint_state_publisher.

This will be moved to a separate package.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add in the GUI package.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

matthew-reynolds commented on joint_state_publisher/src/joint_state_publisher/__init__.py:
```diff
@@ -134,6 +112,15 @@ def init_urdf(self, robot):
+        use_gui = get_param('use_gui')
+        if use_gui is not None:
+            rospy.logerr('GUI is no longer supported by this package; install and run joint_state_publisher_gui')
+            sys.exit(1)
```
I strongly believe we should approach this the same way we're working on removing the GUI dependencies from `actionlib` in ros/actionlib#157.

That is, rather than failing outright, attempt to minimize the breaking change by searching for and launching `joint_state_publisher_gui` if possible, with a warning telling users about the change. In the case that the `joint_state_publisher_gui` package isn't found, print an error and exit(1) as you've done here.

matthew-reynolds commented on README.md:
```diff
@@ -1,9 +1,35 @@
-See the ROS wiki for API documentation and tutorials.
+Given a URDF (either passed on the command-line or via the `/robot_description` topic), this node
+will continually publish default values for all of the movable joints in the URDF to the `/joint_states` topic.
+If the `use_gui` parameter is True, it will also launch a window where the values of the movable joints can be updated.
```
As implemented, this line is no longer accurate. If you choose to do the forwarding method from my other comment above, that functionality should be explained here.

More to the point, I would suggest adding something about the two packages, as with https://github.com/ros/urdf.

Also, not that important but, as is, it feels like these changes to the README don't belong in this PR since the changes are unrelated to the separation of the packages. With adding text about the new structure and change to `use_gui`, it would make a lot more sense.

clalancette replied:
That was just a leftover that I forgot to remove.  I've now removed this.

matthew-reynolds commented on joint_state_publisher_gui/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+catkin_package()
+
+install(PROGRAMS joint_state_publisher_gui/joint_state_publisher_gui DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})
```
Shouldn't the script be placed in `<package_name>/scripts/<script_name>`, not `<package_name>/<package_name>/<script_name>`?

clalancette replied:
Sure.  Fixed in 1db983d.

matthew-reynolds commented on joint_state_publisher_gui/package.xml:
```diff
@@ -0,0 +1,23 @@
+  <exec_depend>joint_state_publisher</exec_depend>
+  <exec_depend>rospy</exec_depend>
+  <exec_depend>python_qt_binding</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
```
What's the rationale for this ordering? I'm surprised it's not alphabetic, or rospy-first-then-alphabetic as it appears it used to be.

clalancette replied:
I just failed at alphabetizing :).  Fixed in 1db983d.

matthew-reynolds commented on joint_state_publisher_gui/package.xml:
```diff
@@ -0,0 +1,23 @@
+  <exec_depend>joint_state_publisher</exec_depend>
+  <exec_depend>rospy</exec_depend>
+  <exec_depend>python_qt_binding</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
```
`sensor_msgs` isn't needed here, as it's not used directly by any files in this package.

clalancette replied:
Good point, removed in 1db983d.

matthew-reynolds reviewed (commented):
I think you're right to target Kinetic & Melodic, though I think we'd be better off with a smoother migration as described in my comment. Then for Noetic we can drop the forwarding altogether.

Couple other comments:

Since you're restructuring and creating a bunch of new files, is this a good opportunity to add license header comments? It appears the other former `robot_model` packages all have the expected 3-clause BSD license w/ Willow Garage as the copyright holder.

Don't think the `screenshot.png` needs to be copied over to the new package.

clalancette committed:
Fixes from review, including new LICENSE files.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
> I think you're right to target Kinetic & Melodic, though I think we'd be better off with a smoother migration as described in my comment. Then for Noetic we can drop the forwarding altogether.
> That is, rather than failing outright, attempt to minimize the breaking change by searching for and launching joint_state_publisher_gui if possible, with a warning telling users about the change. In the case that the joint_state_publisher_gui package isn't found, print an error and exit(1) as you've done here.

Stuck between a rock and a hard place here.  If we do this, then by all rights `joint_state_publisher` should have a dependency on `joint_state_publisher_gui`.  Since `joint_state_publisher_gui` already has a dependency on `joint_state_publisher`, we end up with a circular dependency.  On the other hand, if we don't do this, then we end up with potentially surprising results as you point out.  I guess if we have a graceful fallback option, we can just live with the incorrect dependency hierarchy.  I can take a look at doing that after the holidays.

> 
> Couple other comments:
> 
> Since you're restructuring and creating a bunch of new files, is this a good opportunity to add license header comments? It appears the other former `robot_model` packages all have the expected 3-clause BSD license w/ Willow Garage as the copyright holder.

Good point.  I've added the license stuff in 1db983d.

> 
> Don't think the `screenshot.png` needs to be copied over to the new package.

Actually, I think it is the other way around; `joint_state_publisher` itself no longer needs it (since it is a screenshot of the GUI).  So I've moved the screenshot over to the GUI package.

Thanks for the review.

matthew-reynolds commented:
>I guess if we have a graceful fallback option, we can just live with the incorrect dependency hierarchy

Yeah that's pretty much what I'm suggesting. Rather than an `exec_depend` on `joint_state_publisher_gui`, we have a sort of "soft dependency" where we instruct the user to install the package. Definitely not a clean dependency hierarchy, but it smooths out the migration. Especially since I suspect the "soft dependency" will resolve for most users, since if they're using the GUI they're likely on a `desktop` metapackage install anyways.

> Actually, I think it is the other way around

Good call 👍

clalancette committed:
Reinstate the use_gui option.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
All right, I've now updated it so that `joint_state_publisher` has a "soft" dependency on `joint_state_publisher_gui`, and will attempt to launch the latter if `use_gui` is set to True.  This mimics backwards-compatible behavior with what exists in Kinetic and Melodic.  Assuming we merge this PR, I intend to make a new `noetic` branch where we remove this workaround.

@sloretz @matthew-reynolds Ready for another round of review.

matthew-reynolds commented on joint_state_publisher/scripts/joint_state_publisher:
```diff
@@ -32,6 +32,12 @@
+import sys
+import threading
+
+from python_qt_binding.QtWidgets import QApplication
```
This import pulls back in a pyqt/pyside dependency. Could we move the import to within the `try/except`?

Alternatively, could refactor a bit to add helper functions to the `joint_state_publisher_gui` module to construct the `QApplication` singleton and to call `sys.exit(QApplication.instance()._exec())`, to pull the import out of `jsp` entirely, but that might be adding unnecessary complexity for a workaround that will be removed in Noetic anyways.

matthew-reynolds replied:
On second thought, it might be worth pulling all four of these imports into that `if use_gui:` or `try:` block, so that there aren't unused imports slowing things down when using the `jsp` in intended headless mode. Seems unlikely to make an appreciable difference, but I thought I'd flag it for your consideration.

clalancette replied:
Absolutely, that's a good call.  I'll move them all, it will make it easier to remove the whole thing for Noetic anyway.

clalancette replied:
Done in f1ed8c9

clalancette committed:
Fixes from review.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

matthew-reynolds commented on joint_state_publisher/package.xml:
```diff
@@ -15,9 +15,9 @@
   <url type="bugtracker">https://github.com/ros/joint_state_publisher/issues</url>
 
   <buildtool_depend>catkin</buildtool_depend>
+  <buildtool_depend>python-catkin-pkg</buildtool_depend>
```
Why do we need a `buildtool_depend` on this?

clalancette replied:
As I understand it, we need this dependency since we do:

```
from catkin_pkg.python_setup import generate_distutils_setup
```

In the setup.py.

This comment led me to realize that I had left it out of the package.xml for jsp_gui, which I've rectified in 61dcd3c

matthew-reynolds replied:
I haven't seen that before, in the [catkin package.xml wiki](http://wiki.ros.org/catkin/package.xml) or in the [rospy makefile tutorial](http://wiki.ros.org/rospy_tutorials/Tutorials/Makefile).

It looks like `python-catkin-pkg` is a [dependency of `catkin`](https://github.com/ros/catkin/blob/kinetic-devel/package.xml), so we don't need to explicitly depend on it.

A quick github search for `python-catkin-pkg` in `package.xml` files turns up [a ton of `catkin`'s package.xml](https://github.com/search?utf8=%E2%9C%93&q=python-catkin-pkg+filename%3Apackage.xml&type=Code&ref=advsearch) in people's build spaces, but when I [filtered out `catkin`](https://github.com/search?utf8=%E2%9C%93&q=python-catkin-pkg+NOT+infrastructure+filename%3Apackage.xml&type=Code&ref=advsearch), the only packages I recognized were [`rosjava/genjava`](https://github.com/rosjava/genjava/blob/kinetic/package.xml) and [`rocon_tools/rocon_python_utils`](https://github.com/robotics-in-concert/rocon_tools/blob/devel/rocon_python_utils/package.xml). Certainly doesn't seem to be a common pattern to explicitly depend on it.

clalancette replied:
I'm certainly not a catkin expert.  But at least one of the packages I maintain uses this pattern: https://github.com/ros/kdl_parser/blob/748357e6913408679afbb8bf1adf7d81090aaa76/kdl_parser_py/package.xml#L27 .  (The [urdf_parser_py](https://github.com/ros/urdf_parser_py) package also used to use this pattern, though that was removed in Melodic for reasons of which I'm uncertain)

In the end, I don't feel super strongly about it, though I usually like to explicitly include my direct dependencies.  That way if one of my other dependencies removes a dependency, I don't need to necessarily change my own code.

Maybe @sloretz can weigh in on this one.

sloretz replied:
> It looks like python-catkin-pkg is a dependency of catkin, so we don't need to explicitly depend on it.

Usually I'd advocate for including what we use directly, but in this case I agree that it can be skipped since the import in `setup.py` feels to me like [boilerplate required by catkin](http://docs.ros.org/melodic/api/catkin/html/howto/format2/installing_python.html).

If it's kept, then it should be [`python-catkin-pkg-modules`](https://github.com/ros/rosdistro/blob/70b4b0b3a41a46bc28688efb79436e46f54072ad/rosdep/python.yaml#L856) since it is just the modules that are imported, and not the command line tools. This is important because `python-catkin-pkg-modules` and `python3-catkin-pkg-modules` are side-by-side installable, while `python-catkin-pkg` and `python3-catkin-pkg` are not.

matthew-reynolds replied:
Interesting, must have missed that in my search. I only looked through the first 5-6 pages though, probably just needed to dig further.

Normally I would agree about explicit dependencies, but I suppose I was viewing the package as a core component of `catkin` and not worth explicitly depending upon. There's also the py2 vs py3 logic which should probably be pulled over to this file if we're explicitly depending on it rather than letting the `catkin` package handle that logic.

Definitely would appreciate @sloretz's input.

---

Edit: Whoops, guess I should have refreshed before hitting send :)

clalancette replied:
All right, so it sounds like the consensus is to remove it (particularly in light of the upcoming noetic python3 transition).  I'll go ahead and do that.

matthew-reynolds reviewed (commented):
Thanks for that change, looks great. `import sys` could be moved within the `if use_gui:` block, but then we've got imports in 3 different places, so I like what you've done by leaving it at the top.

Found one last question, see below.

clalancette committed:
Add <buildtool_depend>python-catkin-pkg for jsp_gui.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Remove build_depend on python-catkin-pkg.

We get it implicitly through the catkin dependency.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Honor the num_rows variable in joint_state_publisher_gui.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

sloretz commented on joint_state_publisher/README.md:
```diff
@@ -0,0 +1,34 @@
+* `use_smallest_joint_limits` (bool) - Whether to honor `<safety_controller>` tags in the URDF.  Defaults to True.
+* `source_list` (array of strings) - Each string in this array represents a topic name.  For each string, create a subscription to the named topic of type `sensor_msgs/JointStates`.  Publication to that topic will update the joints named in the message.  Defaults to an empty array.
+* `zeros` (dictionary of string -> float) - A dictionary of joint_names to initial starting values for the joint.  Defaults to an empty dictionary, in which case 0.0 is assumed as the zero for all joints.
+* `dependent_joints` (dictionary of string -> dictionary of 'parent', 'factor', 'offset') - A dictionary of joint_names to the joints that they mimic; compare to the '<mimic>' tag in URDF.  A joint listed here will mimic the movements of the 'parent' joint, subject to the 'factor' and 'offset' provided.  The 'parent' name must be provided, while the 'factor' and 'offset' parameters are optional ( ...
```
In the preview, `'<mimic>'` renders as `''`. Use backticks instead?

clalancette replied:
Will try to fix that.

clalancette replied:
Fixed in 3e59f5a

sloretz commented on joint_state_publisher/src/joint_state_publisher/__init__.py:
```diff
@@ -196,12 +196,6 @@ def source_cb(self, msg):
-            # Signal instead of directly calling the update_sliders method, to
-            # switch to the QThread.  Note that we do this only once at the end
-            # to avoid lots of overhead switching threads.
-            self.gui.sliderUpdateTrigger.emit()
```
It looks like this might update the GUI slider position every time this receives a joint state message. In the refactor, does anything update the sliders when new data comes in?

clalancette replied:
Absolutely nothing; it doesn't work.  Good catch, I'll try to fix this.

clalancette replied:
Fixed in 2ef311d

sloretz commented on joint_state_publisher_gui/scripts/joint_state_publisher_gui:
```diff
@@ -0,0 +1,62 @@
+                                                                   joint_state_publisher.JointStatePublisher(),
+                                                                   num_rows)
+        jsp_gui.show()
+        jsp_gui.sliderUpdateTrigger.emit()
```
What does triggering the event here do?

clalancette replied:
It causes the initial values of the sliders and text boxes to be displayed.  In theory you could do this more directly in the `JointStatePublisherGui.__init__()` method by calling `update_sliders()` there, but this way ensures that you are on the UI thread (and matches what was there before).

sloretz reviewed (commented):
Partial review, coming back to this later.

clalancette committed:
Try to fix README XML tag.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add the ability to update the GUI based on external events.

This can happen if something is published to the joint_state_publisher
source_cb, so in that case we need to update the GUI.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
@sloretz Thanks for the partial review, I've fixed the issues you've identified.  Ready for complete review/second round of review.

clalancette committed:
Fix one more <mimic> tag.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

sloretz commented on joint_state_publisher_gui/src/joint_state_publisher_gui/__init__.py:
```diff
@@ -0,0 +1,196 @@
+import math
```
Nit, missing license boilerplate comment

clalancette replied:
Fixed in 12bd0e5.

sloretz reviewed (approved):
LGTM with one nit. Thanks for keeping the amount of unrelated cleanup to a minimum. It made diffing/reviewing pretty easy.

clalancette committed:
Add license boilerplate to joint_state_publisher_gui.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
Thanks!  Merging now that the buildfarm is turning over.

matthew-reynolds commented:
@clalancette @sloretz Do you plan on updating REP 142/150 to add `joint_state_publisher_gui` to the `desktop` metapackage? Happy to provide a PR if this is your plan.

clalancette commented:
> @clalancette @sloretz Do you plan on updating REP 142/150 to add `joint_state_publisher` to the `desktop` metapackage? Happy to provide a PR if this is your plan.

~~As far as I can tell, REP-142 doesn't need to be updated.  It doesn't include `joint_state_publisher` in the robot metapackage, and I don't think we should change that this late in its lifetime.~~

~~REP-150 is obviously out-of-sync with what is in https://github.com/ros/metapackages; I'll submit a PR to update that soon.~~

Arg, never mind, I was confused.  I forgot about the soft dependency thing.

So it is still the case that REP-142 doesn't have `joint_state_publisher` at all, so I don't think we need to do anything there.  REP-150 should indeed be updated to add `joint_state_publisher_gui` to the desktop metapackage; I'll open a PR for that soon.

clalancette commented:
https://github.com/ros/metapackages/pull/30 and https://github.com/ros-infrastructure/rep/pull/230

matthew-reynolds commented:
>REP-142 doesn't have `joint_state_publisher` at all, so I don't think we need to do anything there

`jsp` is pulled in via `robot_model` (in the robot metapackage) and `urdf_tutorial` (in the desktop_full metapackage). Seems risky to not add `jsp_gui` to `desktop` since previously every install from `robot` upwards contained the gui.

>REP-150 should indeed be updated to add joint_state_publisher_gui to the desktop metapackage

:+1:

matthew-reynolds commented:
I suppose we should also give a scan through the dependant packages on `jsp` and update them accordingly (ie. `urdf_tutorial` uses the jsp gui, but depends only on `jsp`). I will get started on this.

gavanderhoorn commented:
We're getting some questions on ROS Answers about this split ([344992](https://answers.ros.org/question/344992) fi).

Should this change be announced on Discourse to reduce the surprise?

clalancette commented:
@gavanderhoorn Thanks for fielding those ROS answers queries.

I'll go ahead and make an announcement on Discourse.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/split-of-joint-state-publisher-and-joint-state-publisher-gui/12997/1

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/fast-forward-merging-rosbag2-master-api-to-foxy/18927/15
